article.md

```diff
@@ -0,0 +1,333 @@
+    }
+    domain_sep.write(domain_separator);
+    return();
+}
+ ```
+An account deploying mapping contract should perform set_domain_separator within the same transaction as deployment, using multicall.
+### Recreating the message hash
+Since our example is pretty easy, our structure contains only one field (myStarkNetAddress), recreating the hashable string is pretty easy; we just need to concatenate the structure hash with our value.
+
+
+ ```cairo
+ func get_hash{range_check_ptr,  keccak_ptr: felt*, bitwise_ptr : BitwiseBuiltin*}(starknet_address: felt,
+ domain_hash : Uint256) -> (hashed_msg : Uint256)
+{
+    let (starknet_address_h, starknet_address_l) = split_felt(starknet_address);
+    
+    let (encoded_data : Uint256*) = alloc();
+    assert encoded_data[0] = Uint256(TYPE_HASH_LOW, TYPE_HASH_HIGH);
+    assert encoded_data[1] = Uint256(starknet_address_l, starknet_address_h);
+    let (data_hash) = keccak_uint256s_bigend{keccak_ptr=keccak_ptr}(2, encoded_data);
+ ```
+Because `keccak_uint256s_bigend` takes an array of Uint256 structures as an argument, we need to convert our input to this format. <br/><br/>
+The last remaining thing is to combine domain hash and message hash and apply keccak256 to it. There is a small problem — the 2 byte prefix, which prevents us keccak_uint256s_bigend, we need to add the prefix first.
+```cairo
+// value has to be a 16 byte word
+// prefix length = PREFIX_BITS
+func add_prefix{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(value : felt, prefix : felt) -> (
+        result : felt, overflow: felt){
+    let shifted_prefix = prefix * 2 ** 128;
+    // with_prefix is 18 bytes long
+    let with_prefix = shifted_prefix + value;
+    let overflow_mask = 2 ** 16 - 1;
+    let (overflow) = bitwise_and(with_prefix, overflow_mask);
+    let result = (with_prefix - overflow) / 2 ** 16;
+    return (result, overflow);
+}
+```
+We're again operating on 128 bit values, so we need to shift everything to fit our prefix.
+
+```cairo
+    let prefix = PREFIX;
+    let (w0, prefix) = add_prefix(domain_hash.high, prefix);
+    let (w1, prefix) = add_prefix(domain_hash.low, prefix);
+    let (w2, prefix) = add_prefix(data_hash.high, prefix);
+    let (w3, overflow) = add_prefix(data_hash.low, prefix);
+```
+
+We can't use the same high-level keccak implementation as before (`keccak_uint256s_bigend`), since it assumes that our input in bytes is a multiple of 32, which it is not. Because of the 2 overflow bytes, our input is 66 bytes long, and thus `keccak_uint256s_bigend` will assume that we're also counting all the leading zeros of the last value, making our input 96 bytes long, which would result in incorrect hashes. The solution is to use the low-end keccak function - `keccak_bigend`, which unfortunately takes the input in the little-endian format, forcing us to change our variables to it. Note that the `bigend` in the name of the function refers to the output format, not the input format.
```

Comment:
+1 for adding the note of the result, I was just thinking about it when reading this :D